refactor: unify group chat schema

tcm390 opened the pull request:
## Summary

This PR consolidates the two separate group chat systems (user-created groups and NPC-managed groups) into a single unified schema. Previously, we had fragmented tables and inconsistent flows that were difficult to maintain and scale.

## Background

### The Problem: Two Parallel Systems

We had **two completely separate group systems** that evolved independently:

1. **User-Created Groups** (`UserGroup`, `UserGroupMember`, `UserGroupAdmin`, `UserGroupInvite`)
   - Created via the Groups API
   - Had invite flow for adding members
   - Stored admin roles in a separate table

2. **NPC-Managed Groups** (`GroupChatMembership`, `Chat.npcAdminId`)
   - No group table
   - Created by NPCs during game ticks
   - No invite flow - users were added directly
   - Quality tracking (qualityScore, messageCount) for kick mechanics
   - Used `Chat.npcAdminId` instead of a proper Group reference

### Issues with the Old System

- **Two admin systems**: `UserGroupAdmin` table + `ChatAdmin` (dead code, never used)
- **Inconsistent membership tracking**: `UserGroupMember` vs `GroupChatMembership` with different fields
- **Heuristic-based group type detection**: Admin panel used `chat.name.includes(npc.name)` to guess if a group was NPC-managed
- **User-created groups had validation issues**: Members weren't properly validated, leading to bugs (temporary fix in PR #665: https://github.com/BabylonSocial/babylon/pull/665)
- **Duplicate logic**: When a user left a chat, code had to update both `ChatParticipant` AND the appropriate membership table
- **Inconsistent invite behavior**: Adding members during group creation directly added them without invite/consent, while adding members after creation required invite acceptance

### Decision to Refactor

Rather than continuing to patch the fragmented system, I decided to consolidate everything into a unified schema that:
- Supports all group types (`user`, `npc`, `agent`) with the same tables
- Has a single invite flow for consistency
- Uses explicit `group.type` instead of heuristics
- Scales better and is easier to maintain

## Changes

### New Unified Schema

```
Group
├── id, name, description
├── type: 'user' | 'npc' | 'agent'
├── ownerId (who controls it)
├── createdById, createdAt, updatedAt

GroupMember
├── groupId, userId
├── role: 'owner' | 'admin' | 'member'
├── qualityScore, messageCount, lastMessageAt (for NPC kick mechanics)
├── isActive, joinedAt, addedBy
├── kickedAt, kickReason

GroupInvite
├── groupId, invitedUserId, invitedBy
├── status: 'pending' | 'accepted' | 'declined'
├── message, invitedAt, respondedAt

Chat.groupId → Group.id (replaces npcAdminId)
```

## Testing

- Tested NPC group formation by temporarily setting `FORM_NEW_GROUP_CHANCE` to 100%
- Verified NPCs could invite users and users could accept/decline invites
- Verified NPC message posting works correctly with the new schema
- Tested group limit enforcement (max 5 NPC groups per user)

## Screenshot


https://github.com/user-attachments/assets/629b7c5c-5eed-4423-a4da-087f0e39f6ee


https://github.com/user-attachments/assets/53aa8780-c763-4767-9f09-d5cbde5cd31c



## Migration Notes

- **No data migration script included** - schema changes will be applied manually
- Old tables are removed from schema but may still exist in database until migration runs
- `Chat.npcAdminId` column removed - groups now use `Chat.groupId` → `Group.id`

## Known Issues

There are some UI issues related to the refactored group system that will be addressed in a follow-up PR.

## Removed Unused Code

- `apps/web/src/app/api/user-groups/` - entire directory (8 route files)
- `apps/web/src/components/groups/` - `GroupDetailsModal`, `GroupInviteNotification`, `GroupInvitesPanel`, `GroupSettingsModal`, `UserGroupsList`


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Unified group system with owner/admin/member roles, tighter group–chat linking, invite flow updates, group-scoped notifications, and NPC membership limits.

* **Bug Fixes**
  * More reliable membership checks and stronger owner/admin protections for join/leave/promote/demote flows.

* **Refactor**
  * Consolidated invites and membership into a single group-based model; invite/accept flows now return richer info.

* **Breaking Changes**
  * Legacy user-group APIs and several group UI components removed or replaced.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tcm390 committed:
refactor: unify group chat schema

Consolidate UserGroup, GroupChatMembership, and related tables into unified Group, GroupMember, and GroupInvite tables. All group types (user, npc, agent) now use the same schema with Chat.groupId linking to Group.

tcm390 committed:
fix: remove chatDetails from useEffect deps to prevent infinite loop

tcm390 committed:
fix: remove undefined experiencePlugin from runtime plugins array

tcm390 committed:
clean code

tcm390 committed:
fix: handle inactive members on invite accept and auto-add agents

- Accept invite now reactivates inactive members instead of failing on unique constraint
- Inviting an agent (isAgent) auto-adds them without requiring accept

tcm390 committed:
prettier

tcm390 committed:
feat: add group limit check, toast notifications, and fix NPC group posting

- Add env-controlled MAX_ACTIVE_USER_GROUPS (default 5) in GROUP_CONFIG
- Only count NPC groups toward the 5-group limit (user groups excluded)
- Add group limit check when accepting invites (API + MCP handlers)
- Fix GroupInviteCard to show toast notifications instead of error page
- Fix NPC postGroupMessages to use StaticDataRegistry for NPC lookup
- Add 'Already have an open position' to expected trade failures

tcm390 committed:
fix: send invitations to initial group members instead of adding directly

- Human users now receive invitations when added during group creation
- Agents are still auto-added (they can't accept invites)
- Creator remains directly added as owner

tcm390 committed:
fix: handle accepted invite status when re-inviting member

tcm390 committed:
clean code

tcm390 committed:
clean code

tcm390 committed:
clean code

tcm390 committed:
fix: query optimizations, pgEnum, deletion order, and test file updates

- Add pgEnum for group type (user/npc/agent) constraint
- Optimize isInChat query with single join
- Fix group deletion order (chat cleanup before group)
- Fix NPC-specific membership check in ActorSocialActions
- Fix accepted invite status handling in NPC dynamics
- Remove legacy groupsByName lookup in admin route
- Update all group-chat integration tests for unified schema

tcm390 committed:
fix: invite cooldown should only check NPC group memberships

User-created groups should not count toward the NPC group invite cooldown.
This fix ensures users who recently joined user-created groups can still
receive invites from NPCs.

slkzgm commented on packages/db/src/schema/messaging.ts:
```diff
@@ -32,12 +35,11 @@ export const chats = pgTable(
     index('Chat_groupId_idx').on(table.groupId),
```
The comment says "one Chat per Group", but schema only has an index on `Chat.groupId`. Should we add a UNIQUE constraint on `Chat.groupId` (NULLs allowed) to enforce the invariant and avoid `findFirst({ where: { groupId } })` becoming non-deterministic?

slkzgm commented on packages/db/src/schema/messaging.ts:
```diff
@@ -225,92 +144,112 @@ export const notifications = pgTable(
+    // 'owner' = creator/controller of the group
+    // 'admin' = can manage members
+    // 'member' = regular participant
+    role: text('role').notNull().default('member'),
```
`GroupMember.role` is a free-form text column. Would we want a `pgEnum` (or CHECK constraint) here to prevent unexpected values from slipping in (esp. since access control depends on it)?

slkzgm commented on packages/db/src/schema/messaging.ts:
```diff
@@ -225,92 +144,112 @@ export const notifications = pgTable(
     invitedUserId: text('invitedUserId').notNull(),
     invitedBy: text('invitedBy').notNull(),
+    // 'pending' | 'accepted' | 'declined'
     status: text('status').notNull().default('pending'),
```
Same question for `GroupInvite.status`: should this be a `pgEnum` / CHECK constraint? Also, do we intend to allow re-invites by updating the existing row (single invite per group/user) or by allowing multiple historical invites (partial-unique on pending)?

slkzgm commented on packages/engine/src/services/alpha-group-invite-service.ts:
```diff
@@ -116,15 +115,16 @@ export class AlphaGroupInviteService {
+      // Check if already in a group managed by this NPC
+      // The GroupChatService.recordInvite will handle duplicates,
+      // but we do a quick check here for efficiency
       const [existingMembership] = await db
```
This membership check looks broader than the comment "managed by this NPC": it currently skips users who are in *any* active group (including user groups). Is the intent instead to check membership in NPC groups owned by `npcId` (join `groups` + filter `groups.ownerId = npcId` / `groups.type = 'npc'`)?

slkzgm commented on packages/engine/src/services/alpha-group-invite-service.ts:
```diff
@@ -162,27 +164,27 @@ export class AlphaGroupInviteService {
       // Check if user is in invite cooldown
```
Invite cooldown uses the latest membership across all groups. If the cooldown is meant for NPC invites only, should this mirror `NPCGroupDynamicsService.filterUsersForInvite` (join `groups` and filter `groups.type = 'npc'`)?

slkzgm commented on packages/engine/src/services/group-chat-service.ts:
```diff
@@ -275,14 +285,45 @@ export class GroupChatService {
         name: chatName,
         isGroup: true,
         gameId: 'realtime',
+        groupId, // Link Chat → Group
```
Question on semantics: `recordInvite` still adds `ChatParticipant` + `GroupMember` immediately (no `GroupInvite`). After this refactor, are NPC invites supposed to require user acceptance? If yes, I’d expect this to create a `GroupInvite` and let the accept endpoint handle membership/participants.

slkzgm commented on packages/engine/src/services/group-chat-service.ts:
```diff
@@ -275,14 +285,45 @@ export class GroupChatService {
+      groupId = existingChat.groupId;
     }
 
     // Check if participant exists
```
Edge case: if a participant exists but `isActive = false`, we currently do nothing (and unique constraint prevents re-insert). Should we reactivate the participant here? Similar re-join handling likely needed for an existing inactive `GroupMember` (unique on groupId+userId).

slkzgm commented on apps/web/src/app/api/admin/group-invite/route.ts:
```diff
@@ -195,12 +240,13 @@ export const POST = withErrorHandling(async (request: NextRequest) => {
     }
 
     // Record membership
```
`db.groupMember.create` can throw on the `(groupId,userId)` unique constraint if the user was previously a member (inactive) or if two invites race. Consider upserting/reactivating an existing `GroupMember` (and `ChatParticipant.isActive`) rather than unconditional create.

Also: `notifyGroupChatInvite` uses the global `db` (not the `asSystem` tx client). Calling it inside the transaction feels risky; would it be safer to run notifications after the transaction commits?

slkzgm commented on apps/web/src/app/api/groups/invites/[inviteId]/accept/route.ts:
```diff
@@ -84,58 +46,119 @@ export const POST = withErrorHandling(
 
-      // Check if user is already a member
-      const existingMember = await db.userGroupMember.findFirst({
+      // Check if user is at the NPC group limit (only NPC groups count toward the limit)
```
This enforces the NPC group cap without checking the invited group’s type. If a user is at the NPC cap but accepts a *user* group invite, should that be allowed? Might need to fetch the invited group and only apply the cap when `group.type === 'npc'`.

slkzgm commented on apps/web/src/app/api/groups/invites/[inviteId]/accept/route.ts:
```diff
@@ -84,58 +46,119 @@ export const POST = withErrorHandling(
 
-      if (chat) {
-        await db.chatParticipant.create({
+      if (existingMember) {
```
When `existingMember.isActive`, code updates the invite then throws. Because this runs inside the `asUser` transaction, that update will be rolled back, leaving the invite pending and effectively impossible to clear. Suggestion: make accept idempotent (mark accepted + ensure participant active + return 200) instead of throwing.

slkzgm commented on apps/web/src/app/api/groups/invites/[inviteId]/accept/route.ts:
```diff
@@ -84,58 +46,119 @@ export const POST = withErrorHandling(
+      await db.groupInvite.update({
         where: { id: inviteId },
         data: {
           status: 'accepted',
```
This marks *all* `group_invite` notifications as read for the user. Should this be scoped to `inviteId` (and/or `groupId`) to avoid clearing unrelated invites?

slkzgm commented on apps/web/src/app/api/groups/invites/[inviteId]/decline/route.ts:
```diff
@@ -83,7 +46,7 @@ export const POST = withErrorHandling(
+      await db.groupInvite.update({
         where: { id: inviteId },
         data: {
           status: 'declined',
```
Same concern as accept: `updateMany` marks all `group_invite` notifications read. Should this filter by the specific `inviteId`?

slkzgm commented on packages/mcp/src/handlers/tool-handlers.ts:
```diff
@@ -1390,6 +1415,31 @@ export async function executeCreateGroup(
+        role: 'owner',
+        addedBy: agent.userId,
+      },
+      ...args.memberIds.map((memberId, idx) => ({
```
`args.memberIds` isn’t normalized (can include duplicates and/or `agent.userId`). That will violate unique constraints on `ChatParticipant(chatId,userId)` and `GroupMember(groupId,userId)` for `createMany`. Should we dedupe + filter before inserting?

slkzgm commented on packages/mcp/src/handlers/tool-handlers.ts:
```diff
@@ -1475,65 +1548,139 @@ export async function executeMarkNotificationsRead(
+      const group = groupsMap.get(invite.groupId);
+      return {
+        id: invite.id,
+        groupId: chatsMap.get(invite.groupId) || invite.groupId, // Return chatId for backward compat
```
Returning `chatId` in a field named `groupId` ("for backward compat") seems risky for downstream consumers. Could we return both `groupId` and `chatId` (and keep `groupId` truthful)?

slkzgm commented on packages/mcp/src/handlers/tool-handlers.ts:
```diff
@@ -1475,65 +1548,139 @@ export async function executeMarkNotificationsRead(
   args: AcceptGroupInviteArgs
 ): Promise<AcceptGroupInviteResult> {
-  const invite = await db.chatInvite.findUnique({
+  const invite = await db.groupInvite.findUnique({
```
Accept flow isn’t idempotent and isn’t transactional: no `invite.status === 'pending'` check, and it `create`s ChatParticipant/GroupMember unconditionally (unique constraints). If any insert fails after the invite update, state can be inconsistent. Suggest wrapping in a transaction and reusing the web route’s reactivation logic.

slkzgm reviewed (changes requested):
Thanks for taking on a big, high-impact refactor here — unifying the group schema is the right direction.

I reviewed the diff against `staging` and also skimmed the existing bot feedback. I think we’re close, but I’m requesting changes because there are a few correctness / consistency issues that look merge-blocking.

**Main blockers (please address before merge)**
1) **AlphaGroupInviteService eligibility checks** look too broad (users in *any* group can get excluded / cooled down for NPC invites).
2) **Invite semantics are inconsistent across the codebase** (some “invites” create membership immediately; others create `GroupInvite` and require acceptance). Please confirm the intended UX for NPC invites and align the implementations.
3) **Idempotency / uniqueness**: several flows can hit `GroupMember(groupId,userId)` and `ChatParticipant(chatId,userId)` unique constraints (notably MCP accept/create and admin NPC invite).
4) **Notifications**: accept/decline currently mark *all* `group_invite` notifications as read (likely too broad).

**Non-blocking / follow-ups**
- Consider DB-level constraints for `GroupMember.role` and `GroupInvite.status` (enum / CHECK), and enforcing the “one chat per group” invariant.
- Migration plan: PR description says there’s no migration script. I’d strongly prefer we ship a Drizzle migration (or at least a documented, repeatable migration procedure) so deploys don’t rely on manual steps.

Verdict: **Request changes** (see inline comments for specifics).

SYMBaiEX committed:
feat: enhance group invite and membership handling

- Implement NPC group limit check when inviting users to groups.
- Refactor membership recording to handle existing members by reactivating them instead of creating duplicates.
- Update group invite logic to send notifications to invited users.
- Ensure proper cleanup of group members and invites during group deletion.
- Introduce unified schema for group and chat management, improving data consistency.

SYMBaiEX committed:
refactor: clean up code formatting and re-export SSEMessage type for consistency

- Adjust formatting in `npc-group-dynamics-service.ts` for improved readability.
- Re-export `SSEMessage` type in `useSSE.ts` for backwards compatibility.
- Update group chat creation calls in integration tests for better code clarity.

SYMBaiEX committed:
refactor: update group invite notification logic and improve invite handling

- Modify `notifyGroupChatInvite` to accept `groupId` and `inviteId` for better notification linking.
- Adjust group invite handling to ensure proper notifications are sent when users are invited or re-invited.
- Enhance comments for clarity regarding notification processes and invite status management.

SYMBaiEX committed:
fix: remove kickedAt/kickReason from chatParticipant updates (schema refactor)

SYMBaiEX committed:
fix: address CodeRabbit review feedback

- Replace Actor table query with StaticDataRegistry in integration-a2a-sdk.ts
- Update pre-dev-local.ts to check User table instead of dropped Actor table
- Remove duplicate chat deletion in test cleanup functions
- Optimize NPC group count query in admin/group-invite using count()

SYMBaiEX committed:
refactor: improve group invite and membership handling with deterministic IDs

- Introduce a deterministic group ID generation method to ensure idempotency when creating groups from chat IDs.
- Refactor group invite logic to utilize atomic upserts, preventing race conditions during group and chat creation.
- Enhance membership handling by using raw SQL for efficient updates and ensuring proper activation of inactive members.
- Update notification logic to include group ID and invite ID for better tracking of invites.

SYMBaiEX committed:
chore: remove obsolete migration files and clean up migration metadata

- Deleted migration file `0011_fix_group_member_partial_unique.sql` as it is no longer needed.
- Removed snapshot files `0005_snapshot.json` and `0006_snapshot.json` to streamline migration history.
- Cleaned up `_journal.json` to reflect the removal of the deleted migration files.

SYMBaiEX committed:
fix: improve console log formatting in post-push-indexes script

- Adjusted console log statement for better readability by formatting it across multiple lines.

SYMBaiEX committed:
fix: correct import path for StaticDataRegistry in integration-a2a-sdk
